providers/implementations/exchange/kdf_exch.c

```diff
@@ -92,16 +94,33 @@ static int kdf_derive(void *vpkdfctx, unsigned char *secret, size_t *secretlen,
                       size_t outlen)
 {
     PROV_KDF_CTX *pkdfctx = (PROV_KDF_CTX *)vpkdfctx;
+    size_t kdfsize;
+    int ret;
 
     if (!ossl_prov_is_running())
         return 0;
 
+    kdfsize = EVP_KDF_CTX_get_kdf_size(pkdfctx->kdfctx);
+
     if (secret == NULL) {
-        *secretlen = EVP_KDF_CTX_get_kdf_size(pkdfctx->kdfctx);
+        *secretlen = kdfsize;
         return 1;
     }
 
-    return EVP_KDF_derive(pkdfctx->kdfctx, secret, outlen, NULL);
+    if (kdfsize != SIZE_MAX) {
```

Comment:
This looks odd. Why is this necessary?

Reply:
Because depending on the mode (EXTRACT_AND_EXPAND, EXTRACT_ONLY or EXPAND_ONLY), the output size may be fixed or indefinite.  Kinda documented here:

https://github.com/openssl/openssl/blob/d2399d8cd29f56e6614f0b3db4e7e563a745902a/doc/man3/EVP_KDF.pod?plain=1#L157-L160